README
About
=====

GNU libextractor is a library used to extract meta data from files of
arbitrary type.  It is designed to use helper-libraries to perform the
actual extraction, and to be trivially extendable by linking against
external extractors for additional file types.  Common use-cases for
GNU libextractor include detail-views in file managers, detailed
search results in file-sharing networks and general information
gathering in forensics investigations and penetration testing.

GNU libextractor is a simple C library with a small API.  Bindings for
GNU libextractor exists for many languages in addition to the standard
C/C++ API (we know about bindings for Java, Perl, PHP, Mono, Python,
Ruby).

libextractor uses a plugin mechanism to enable developers to quickly
add extractors for additional formats.  Plugins are executed
out-of-process and can thus bugs in them (or the libraries that they
use) cannot crash the main application.  libextractor typically ships
with a few dozen plugins that can be used to obtain keywords from
common file types.

More detailed documentation is available in the GNU libextractor
manual.  libextractor is an official GNU package and available from
http://www.gnu.org/s/libextractor/.


extract
=======

extract is a simple command-line interface to GNU libextractor.


Dependencies
============

* GNU C/C++ compiler
* libltdl 2.2.x (from GNU libtool)
* GNU libtool 2.2 or higher
* GNU gettext

The following dependencies are all optional, but should be
available in order for maximum coverage:

* libarchive
* libavutil / libavformat / libavcodec / libswscale (ffmpeg)
* libbz2 (bzip2)
* libexiv2
* libflac
* libgif (giflib)
* libglib (glib)
* libgtk+
* libgsf
* libgstreamer
* libjpeg (v8 or later)
* libmagic (file)
* libmpeg2
* libmp4v2
* librpm
* libsmf
* libtidy
* libtiff
* libvorbis / libogg
* libz (zlib)

When building libextractor binaries, please make sure all of these
dependencies are available and configure detects a sufficiently recent
installation.  Otherwise the build system may automatically build only
a subset of GNU libextractor resulting in mediocre meta data
production.

Finally, 'zzuf' is a fuzzing tool that can optionally be detected by
the build system and be used for debugging / testing.  It is not required
at runtime or for normal builds.



NEWS
Tue Sep 25 16:25:05 CEST 2012
        Releasing GNU libextractor 1.0.0. 

Fri Aug 17 21:28:33 CEST 2012
    	Major changes to plugin mechanism now allow out-of-process
	plugins full random-access to the entire file.  Most plugins
	have now been rewritten to the new plugin API (0.7).  The external
	(libextractor) API will remain unchanged and compatible with LE 0.6.

	As part of the rewrite, many plugins were changed to use 
	standard 3rd party libraries (libjpeg, libtiff, libgif, 
	libtidy, libmagic) for parsing.  A new plugin based on
	gstreamer replaces many existing multimedia plugins.

	Automated testcases for (almost all) of the plugins
	were also written and the documentation was updated.

	The license was changed to GNU GPL v3+.  
	PDF support removed (GPLv2-only).
	REAL support removed (not yet ported).

Mon Nov 28 12:17:42 CET 2011
    	Added MKV (Matroska) plugin.
	Releasing libextractor 0.6.3.

Sun Jun 13 13:15:43 CEST 2010
	Releasing libextractor 0.6.2.

Sun Mar 14 00:09:45 CET 2010
	Releasing libextractor 0.6.1.

Wed Jan 13 17:11:07 CET 2010
        Major API change (no longer returning linked lists, using callbacks instead).
	Releasing libextractor 0.6.0.

Sat Jul  4 23:05:22 CEST 2009
	Releasing libextractor 0.5.23.

Fri Feb 20 00:28:00 MST 2009
	Releasing libextractor 0.5.22.

Sun Nov  2 20:19:02 MST 2008
	Releasing libextractor 0.5.21.

Tue Aug 12 04:40:49 EEST 2008
	Added an S3M (Scream Tracker 3 Module) plugin.
	Added an XM (eXtended Module) plugin.
	Added an IT (Impulse Tracker) plugin.

Mon Jun 23 19:05:07 EET 2008
	Fixed concurrency issues in plugin (un-)loading by adding locking
	around libltdl functions.

Fri Jun 20 23:34:02 EET 2008
	Added an FFmpeg-based thumbnail extractor plugin, initially
	supporting only bmp and png files.

Mon Mar 21 00:00:52 MDT 2008
	Releasing libextractor 0.5.20a.

Thu Mar 20 23:38:47 MDT 2008
	Releasing libextractor 0.5.20.

Wed Feb 13 10:16:55 EET 2008
	Added a plugin for AppleSingle/AppleDouble files.

Sat Jan 12 14:14:32 MST 2008
	Releasing libextractor 0.5.19a.

Mon Jan  7 08:46:40 MST 2008
	Releasing libextractor 0.5.19.

Wed Dec 26 19:38:22 MST 2007
	Added a FLAC (.flac) plugin.

Wed Dec 26 14:50:11 EET 2007
	Added a Flash Video (.flv) plugin.

Sat Jan  6 14:27:18 EET 2007
	Added an NSFE (Extended Nintendo Sound Format) plugin.

Mon Jan  1 19:10:40 MST 2007
	Added pkgconfig support.  Releasing libextractor 0.5.17.

Mon Nov 20 22:08:55 EET 2006
	Added an SID (C64 music file) plugin.

Sat Nov 11 00:04:34 EET 2006
	Added an NSF (NES Sound Format) plugin.

Tue Apr 18 14:44:37 PDT 2006
	Added dictionaries for Finnish, French, Gaelic and Swedish
	(for printable extractors).

Thu Mar  9 17:55:09 PST 2006
	Word history extraction works (wordleaker).

Sat Feb 18 17:39:10 PST 2006
	Releasing libextractor 0.5.10.

Fri Dec 23 11:28:23 PST 2005
	Releasing libextractor 0.5.9.

Thu Jul 14 22:30:45 CEST 2005
	exiv2 works.

Sat May 21 13:42:46 EST 2005
	Releasing libextractor 0.5.0.

Fri May  6 06:02:02 EST 2005
        Added Python binding.

Thu Mar 17 08:07:04 EST 2005
	libextractor becomes a GNU package.

Thu Feb 24 00:34:15 EST 2005
	Thumbnails work.

Sun Feb 20 16:36:17 EST 2005
	Helix works (new Real format).

Sat Nov 13 13:23:23 EST 2004
	Releasing libextractor 0.3.11.

Sat Oct 23 13:21:23 EST 2004
	Releasing libextractor 0.3.10.

Fri Oct 22 19:18:38 EST 2004
	id3v2.3 and id3v2.4 work.

Sun Oct 17 18:12:11 EST 2004
	tar and tar.gz work.

Sun Oct 17 17:42:16 EST 2004
	deb works.

Sun Oct 17 13:52:25 EST 2004
	man works.

Sat Oct  2 18:02:56 EST 2004
	dvi works.

Thu Sep 23 11:25:42 EST 2004
	Added support for StarOffice and ID3v2.
	
Fri Sep 10 22:00:09 EST 2004
	Added support for RipeMD-160.

Fri Sep 10 19:49:39 EST 2004
	Added support for SHA-1 and MD5.

Fri Sep 10 10:35:27 EST 2004
	OpenOffice works.

Mon Aug 30 23:16:17 IST 2004
	ole2 works (WinWord, Excel, PowerPoint).

Wed Apr 28 19:28:39 EST 2004
	Releasing libextractor 0.3.1.

Wed Aug 25 18:59:13 IST 2004
	GNU gettext support added.

Wed Apr 28 01:26:53 EST 2004
	elf works.

Sat Apr 10 01:34:04 EST 2004
	riff / avi works.

Sat Apr 10 00:30:19 EST 2004
	mpeg works.

Sun Apr  4 20:24:39 EST 2004
	tiff works.

Wed Jul 16 13:41:34 EST 2003
	Releasing 0.2.5.

Mon Jun 30 21:27:42 EST 2003
	Releasing 0.2.4.

Thu Jan 23 00:34:20 EST 2003
	rpm works.

Tue Jan  7 18:31:41 CET 2003
	ps works.

Tue Dec 31 15:26:00 EST 2002
	pdf works.

Tue Dec 17 20:36:13 CET 2002
	MIME-type detection works.

Fri Jun  7 01:48:34 EST 2002
	real meta formats (rm, ra) work.

Tue Jun  4 23:21:38 EST 2002
	png works.

Sun Jun  2 22:49:17 EST 2002
	jpeg and HTML work. 

Sat May 18 02:35:39 EST 2002
	ogg works.

Thu May 16 00:04:03 EST 2002
	mp3 works.

Wed May 15 21:38:31 EST 2002
	Project started.


src/common/pack.c
/*
Catlib Copyright Notice

The author of this software is Christopher Adam Telfer
Copyright (c) 1998, 1999, 2000, 2001, 2002
by Christopher Adam Telfer.  All Rights Reserved.

Permission to use, copy, modify, and distribute this software for any
purpose without fee is hereby granted, provided that the above copyright
notice, this paragraph, and the following two paragraphs appear in all
copies, modifications, and distributions.

IN NO EVENT SHALL THE AUTHOR BE LIABLE TO ANY PARTY FOR DIRECT, INDIRECT,
SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES, INCLUDING LOST PROFITS,
ARISING OUT OF THE USE OF THIS SOFTWARE AND ITS DOCUMENTATION, EVEN IF
THE AUTHOR HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

THE AUTHOR SPECIFICALLY DISCLAIMS ANY WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
PARTICULAR PURPOSE.  THE SOFTWARE AND ACCOMPANYING DOCUMENTATION, IF
ANY, PROVIDED HEREUNDER IS PROVIDED "AS IS".   THE AUTHOR HAS NO
OBLIGATION TO PROVIDE MAINTENANCE, SUPPORT, UPDATES, ENHANCEMENTS, OR
MODIFICATIONS.

*/

#include "platform.h"
#include "pack.h"

#if !(defined(_WIN32) && defined(cbNDRContext))
typedef unsigned char byte;
#endif
typedef unsigned short half;
typedef unsigned int word;
typedef signed char sbyte;
typedef signed short shalf;
typedef signed int sword;



int
EXTRACTOR_common_cat_unpack (const void *buf, 
			     const char *fmt,
			     ...)
{
  va_list ap;
  word maxlen, len, *wordp;
  void *arr;
  byte *bp, *bytep, *newbuf;
  half *halfp;
  long long *ll;
  sbyte *sbytep;
  shalf *shalfp;
  sword *swordp;
  int npacked;
  unsigned int nreps, i, isnonprefixed = 1;     /* used for 'a' types only */
  struct cat_bvec *cbvp;
  char *cp;

  bp = (byte *) buf;
  npacked = 0;

  va_start (ap, fmt);

  while (*fmt)
    {
      nreps = 1;

      if (isdigit ( (unsigned char) *fmt))
        {
          /* We use cp instead of format to keep the 'const' qualifier of fmt */
          nreps = strtoul (fmt, &cp, 0);
          fmt = cp;
          if (*fmt == 'a')
            isnonprefixed = 0;
        }

      switch (*fmt)
        {
        case 'B':
        case 'b':
          bytep = va_arg (ap, byte *);
          for (i = 0; i < nreps; ++i)
            {
              *bytep = *bp++;
              ++bytep;
              npacked += 1;
            }
          break;



        case 'h':
          halfp = va_arg (ap, half *);
          for (i = 0; i < nreps; ++i)
            {
              *halfp = *bp++;
              *halfp |= *bp++ << 8;
              ++halfp;
              npacked += 2;
            }
          break;

        case 'H':
          halfp = va_arg (ap, half *);
          for (i = 0; i < nreps; ++i)
            {
              *halfp = *bp++ << 8;
              *halfp |= *bp++;
              ++halfp;
              npacked += 2;
            }
          break;


        case 'w':
          wordp = va_arg (ap, word *);
          for (i = 0; i < nreps; ++i)
            {
              *wordp = *bp++;
              *wordp |= *bp++ << 8;
              *wordp |= *bp++ << 16;
              *wordp |= *bp++ << 24;
              ++wordp;
              npacked += 4;
            }
          break;

        case 'x':
          ll = va_arg (ap, long long *);
          for (i = 0; i < nreps; ++i)
            {
              *ll = ((long long) *bp++);
              *ll |= ((long long) *bp++) << 8;
              *ll |= ((long long) *bp++) << 16;
              *ll |= ((long long) *bp++) << 24;
              *ll |= ((long long) *bp++) << 32;
              *ll |= ((long long) *bp++) << 40;
              *ll |= ((long long) *bp++) << 48;
              *ll |= ((long long) *bp++) << 56;
              ++ll;
              npacked += 8;
            }
          break;

        case 'W':
          wordp = va_arg (ap, word *);
          for (i = 0; i < nreps; ++i)
            {
              *wordp = *bp++ << 24;
              *wordp |= *bp++ << 16;
              *wordp |= *bp++ << 8;
              *wordp |= *bp++;
              ++wordp;
              npacked += 4;
            }
          break;

        case 'X':
          ll = va_arg (ap, long long *);
          for (i = 0; i < nreps; ++i)
            {
              *ll = ((long long) *bp++) << 56;
              *ll |= ((long long) *bp++) << 48;
              *ll |= ((long long) *bp++) << 40;
              *ll |= ((long long) *bp++) << 32;
              *ll |= ((long long) *bp++) << 24;
              *ll |= ((long long) *bp++) << 18;
              *ll |= ((long long) *bp++) << 8;
              *ll |= ((long long) *bp++);
              ++ll;
              npacked += 8;
            }
          break;


        case 'A':
          if (isnonprefixed)
            {
              maxlen = va_arg (ap, word);
              arr = va_arg (ap, void *);

              len = *bp++ << 24;
              len |= *bp++ << 16;
              len |= *bp++ << 8;
              len |= *bp++;

              if (len > maxlen)
		{
		  va_end (ap);
		  return -1;
		}

              memmove (arr, bp, len);
              bp += len;

              npacked += len;
            }
          else
            {
              cbvp = va_arg (ap, struct cat_bvec *);
              for (i = 0; i < nreps; ++i)
                {
                  maxlen = cbvp->len;
                  arr = cbvp->data;

                  len = *bp++ << 24;
                  len |= *bp++ << 16;
                  len |= *bp++ << 8;
                  len |= *bp++;

                  if (len > maxlen)
                    return -1;

                  memmove (arr, bp, len);
                  cbvp->len = len;
                  bp += len;

                  ++cbvp;
                  npacked += len;
                }
              isnonprefixed = 1;
            }
          break;

        case 'C':
        case 'c':
          sbytep = va_arg (ap, sbyte *);
          for (i = 0; i < nreps; ++i)
            {
              *sbytep = *bp++;

              if ((sizeof (sbyte) > 1) && (*sbytep & 0x80))
                *sbytep |= (~0) << ((sizeof (sbyte) - 1) * 8);

              ++sbytep;
              npacked += 1;
            }
          break;


        case 's':
          shalfp = va_arg (ap, shalf *);
          for (i = 0; i < nreps; ++i)
            {
              *shalfp = *bp++;
              *shalfp |= *bp++ << 8;

              if ((sizeof (shalf) > 2) && (*shalfp & 0x8000))
                *shalfp |= (~0) << ((sizeof (shalf) - 2) * 8);

              ++shalfp;
              npacked += 2;
            }
          break;

        case 'S':
          shalfp = va_arg (ap, shalf *);
          for (i = 0; i < nreps; ++i)
            {
              *shalfp = *bp++ << 8;
              *shalfp |= *bp++;

              if ((sizeof (shalf) > 2) && (*shalfp & 0x8000))
                *shalfp |= (~0) << ((sizeof (shalf) - 2) * 8);

              ++shalfp;
              npacked += 2;
            }
          break;

        case 'l':
          swordp = va_arg (ap, sword *);
          for (i = 0; i < nreps; ++i)
            {
              *swordp = *bp++;
              *swordp |= *bp++ << 8;
              *swordp |= *bp++ << 16;
              *swordp |= *bp++ << 24;

              if ((sizeof (swordp) > 4) && (*swordp & 0x80000000))
                *swordp |= (~0) << ((sizeof (sword) - 4) * 8);

              ++swordp;
              npacked += 4;
            }
          break;

        case 'L':
          swordp = va_arg (ap, sword *);
          for (i = 0; i < nreps; ++i)
            {
              *swordp = *bp++ << 24;
              *swordp |= *bp++ << 16;
              *swordp |= *bp++ << 8;
              *swordp |= *bp++;

              if ((sizeof (swordp) > 4) && (*swordp & 0x80000000))
                *swordp |= (~0) << ((sizeof (sword) - 4) * 8);

              ++swordp;
              npacked += 4;
            }
          break;

        case 'P':
          cbvp = va_arg (ap, struct cat_bvec *);
          for (i = 0; i < nreps; ++i)
            {
              len = *bp++ << 24;
              len |= *bp++ << 16;
              len |= *bp++ << 8;
              len |= *bp++;

              newbuf = (byte *) malloc (len);

              if (!newbuf)
                {
                  int j;
                  for (j = 0; j < i; j++)
                    free (cbvp[i].data);
		  va_end (ap);
                  return -1;
                }

              memmove (newbuf, bp, len);
              cbvp[i].data = newbuf;
              cbvp[i].len = len;

              bp += len;
              npacked += len;
            }
          break;

        default:
          va_end (ap);
          return -1;
        }

      ++fmt;
    }

  va_end (ap);
  return 0;
}
